src/dms/dmsScanner.cpp

```diff
@@ -0,0 +1,346 @@
+      {
+         _curRID._extent = nextExtentID() ;
+         _firstRun = TRUE ;
+         return SDB_OK ;
+      }
+      return SDB_DMS_EOC ;
+   }
+
+   INT32 _dmsExtScannerBase::advance( dmsRecordID &recordID,
+                                      dmsRecordGenerator &generator,
+                                      pmdEDUCB *cb )
+   {
+      INT32 rc = SDB_OK ;
+
+      if ( _firstRun )
+      {
+         rc = _firstInit( cb ) ;
+         PD_RC_CHECK( rc, PDWARNING, "first init failed, rc: %d", rc ) ;
+      }
+
+      rc = _fetchNext( recordID, generator, cb ) ;
+      if ( rc )
+      {
+         // Do not write error log when EOC.
+         if ( SDB_DMS_EOC != rc )
+         {
+            PD_LOG( PDERROR, "Get next record failed, rc: %d", rc ) ;
+         }
+         goto error ;
+      }
+
+   done:
+      return rc ;
+   error:
+      recordID.reset() ;
+      _curRID._offset = DMS_INVALID_OFFSET ;
+      goto done ; 
+   }
+
+   void _dmsExtScannerBase::stop()
+   {
+      _next = DMS_INVALID_OFFSET ;
+      _curRID._offset = DMS_INVALID_OFFSET ;
+   }
+
+   _dmsExtScanner::_dmsExtScanner( dmsBlockData *su,
+                                   _dmsMBContext *context,
+                                   dmsExtentID curExtentID,
+                                   DMS_ACCESS_TYPE accessType,
+                                   INT64 maxRecords,
+                                   INT64 skipNum,
+                                   INT32 flag )
+   : _dmsExtScannerBase( su, context, curExtentID, accessType,
+                         maxRecords, skipNum, flag )
+   {
+   }
+
+   _dmsExtScanner::~_dmsExtScanner()
+   {
+   }
+
+   INT32 _dmsExtScanner::_firstInit( pmdEDUCB *cb )
+   {
+      INT32 rc          = SDB_OK ;
+
+      _extRW = _pSu->extent2RW( _curRID._extent, _context->mbID() ) ;
+      _extRW.setNothrow( TRUE ) ;
+      _extent = _extRW.readPtr<dmsExtent>() ;
+      if ( NULL == _extent )
+      {
+         rc = SDB_INVALIDARG ;
+         goto error ;
+      }
+/*      if ( cb->isInterrupted() )
+      {
+         rc = SDB_APP_INTERRUPT ;
+         goto error ;
+      }
+*/
+      if ( !_context->isMBLock( _mbLockType ) )
+      {
+         rc = _context->mbLock( _mbLockType ) ;
+         PD_RC_CHECK( rc, PDERROR, "dms mb lock failed, rc: %d", rc ) ;
+      }
+      if ( !_extent->validate( _context->mbID() ) )
+      {
+         rc = SDB_SYS ;
+         goto error ;
+      }
+
+      _cb   = cb ;
+      _next = _extent->_firstRecordOffset ;
+
+      // unset first run
+      _firstRun = FALSE ;
+
+   done:
+      return rc ;
+   error:
+      goto done ;
+   }
+
+   INT32 _dmsExtScanner::_fetchNext( dmsRecordID &recordID,
+                                     dmsRecordGenerator &generator,
+                                     pmdEDUCB *cb )
+   {
+      INT32 rc                = SDB_OK ;
+      ossValuePtr recordDataPtr ;
+      dmsRecordData recordData ;
+
+      while ( DMS_INVALID_OFFSET != _next && 0 != _maxRecords )
```

Comment:
you will be looping till done or finish with maxrecord here. I am wondering how many records you will be returning and will you overwrite previous found record?

Reply:
It will be looping till done, once a valid record id found, it will be put in the generator and return to the caller.